fideslib/db/base_class.py

```diff
@@ -0,0 +1,302 @@
+        else:
+            db_obj = cls.create(db=db, data=data)
+        return db_obj
+
+    @classmethod
+    def get_or_create(
+        cls, db: Session, *, data: dict[str, Any]
+    ) -> tuple[bool, FidesBase | None]:
+        """Fetch an object, or create it if none is found"""
+        db_obj = db.query(cls).filter_by(**data).first()
+        created = False
+        if not db_obj:
+            db_obj = cls.create(db=db, data=data)  # type: ignore
+            created = True
+        return created, db_obj
+
+    @classmethod
+    def update_with_class(
+        cls, db: Session, *, conditions: Any, values: dict[str, Any]
+    ) -> int:
+        """Update all objects within a filter at database level"""
+        return db.query(cls).filter(conditions).update(values=values)
+
+    @classmethod
+    def delete_with_class(
+        cls, db: Session, *, id: str  # pylint: disable=W0622
+    ) -> FidesBase | None:
+        """Delete an existing row from the database from the object's class"""
+        obj = db.query(cls).get(id)
+        if obj is None:
+            return None
+        db.delete(obj)
+        db.commit()
+        return obj
+
+    @classmethod
+    def delete_all(cls, db: Session) -> int:  # pylint: disable=W0622
+        """Delete all rows in this table"""
+        deleted_count = db.query(cls).delete()
+        return deleted_count
+
+    def refresh_from_db(self, db: Session) -> FidesBase:
+        """Returns a current version of this object from the database"""
+        return db.query(self.__class__).get(self.id)
+
+    def update(self, db: Session, *, data: dict[str, Any]) -> FidesBase:
+        """Update specific row with supplied values"""
+        # Set self.key where applicable
+        if hasattr(self, "key") and "key" in data:
+            data["key"] = get_key_from_data(data, self.__class__.__name__)
+
+        # Update dataset with values in data
+        for key, val in data.items():
+            setattr(self, key, val)
+
+        return self.save(db=db)
+
+    def delete(self, db: Session) -> FidesBase | None:
+        """Delete an existing row in the database from an existing object in memory"""
+        db.delete(self)
+        db.commit()
+        return self
+
+    def save(self, db: Session) -> FidesBase:
+        """Save the current object over an existing row in the database"""
+        if hasattr(self, "key"):
+            key = getattr(self, "key", "")
+
+            is_valid = False
+            if key is not None:
+                valid_key_chars = (
+                    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"
+                )
+                to_validate = key
+                for char in valid_key_chars:
+                    to_validate = to_validate.replace(char, "")
+
+                # Any invalid chars won't have been replaced above, so the length of our
+                # validation var will still be >0
+                is_valid = len(to_validate) == 0
```

Comment:
```python
import re

ALLOWED_KEYCHARS = re.compile(r"[A-z0-9\-_]")
```
then:

```suggestion
                is_valid = len(ALLOWED_KEYCHARS.sub("", key)) == 0
```

Alternatively, you can define the regexp inline within this function, but frequent re-use implies that precompiling would be more performant.

Reply:
Nice, I like this!